### DISCUSSION TITLE
Ideas for future development of pvlib iotools

### DISCUSSION THREAD

**Discussion by AdamRJensen** [pvlib/pvlib-python#1528] [General]
I've attempted to provide a bit of an overview of the *possible* development of the iotools and gather some notes on the inconsistencies in how the functions are implemented. The items are primarily based on the existing open pull requests and issues with the `io` tag, which can be found [here](https://github.com/pvlib/pvlib-python/labels/io).

**General**
- [x] iotools landing page / gallery. I think that it would be of great value to have some sort of landing page with an overview of supported file formats and data sources. There is some discussion here: #1304, #1754
- [x] Table with data providers summarizing main differences (this is done in https://doi.org/10.1016/j.solener.2023.112092) and partly in #2447 

**Reanalysis data**
- [x] ERA5 #2573, #1264
- [x] MERRA2 #2572, #1274 
See also #1484
- [x] [NASA POWER](https://power.larc.nasa.gov/) #2500 

**Commercial data**
See #1310 for a general discussion.
- [x] [SolarAnywhere](https://www.solaranywhere.com/) #1497
- [x] [Solcast](https://solcast.com/) #1313 #1875
- [x] [Solargis](https://solargis.com/) #1969
- [x] [Meteonorm](https://meteonorm.com/en/) #2499

**Other data sources**
- [ ] Extend `read_bsrn` to also parse LR1000 etc.
- Horizon data
  - [ ] #1290
  - [ ] #1295
  - [x] #1395
- Rainfall data
  - [ ] #668
  - [x] #1293

**Enhancements / inconsistencies**
- [x] #1517
- [x] #1773
- [x] Add `coerce_year` parameter to `get_pvgis_tmy` #2138
- [x] Make new function ``get_srml`` that is similar to `read_srml_month_from_solardat` but accepts `start`/`end` #1779
- [x] Change `server` to `url` in `pvlib.iotools.get_cams` #2463
- [x] Allow `start`/`end` to be string by wrapping in `pd.to_datetime(start)`, see this [comment](https://github.com/pvlib/pvlib-python/pull/1767#discussion_r1231015136)
- [x] Replacing `read_srml_month_from_solardat` with `get_srml` #1779
- [x] Add `get_solrad` function #1967 
- ~~For all ``get`` functions, remove `timeout` and replace by kwargs that pass to `requests.get`~~
- [ ] #1420
- [ ] Variable mapping of `read_tmy2`
- [ ] Variable mapping of `read_epw`
- [ ] Add a generic `Latitude.from_meta()` function (similar to `from_epw` and `from_tmy`)
- [ ] Address #1156 
- [ ] Renaming of `pvlib.iotools.read_midc_raw_data_from_nrel`
- [ ] Add ``get_surfrad`` function

**Returned parameters**
Another inconsistency is whether the iotools functions return a single data object or a tuple of (data, meta). A single data object is returned when the metadata is not being parsed or the file does not contain any metadata. It is likely that several of these functions might be improved in the future to also parse metadata; however, returning a tuple of (data, meta) would result in a breaking change which is undesirable. Therefore, I suggest that all new functions return a tuple of (data, meta); and in the case that the metadata is not parsed or nonexistent the meta object is simply an empty dictionary. Given the large number of existing functions, I do not suggest that we change the existing functions. The following functions only return a single DataFrame:
- ``read_midc``
- ``read_midc_raw_data_from_nrel``
- ``read_crn``
- ``read_srml``
- [x] ``read_srml_month_from_solardat`` #1779 
- [x] ``read_solrad`` #1968
- [x] ``x_pvgis_hourly`` #2462
- [x] ``x_pvgis_tmy`` #2470

**High-level changes**
- [ ] Use pandas [IntervalIndex](https://pandas.pydata.org/docs/reference/api/pandas.IntervalIndex.html#pandas.IntervalIndex) for the datetime index, instead of `DatetimeIndex` which is labeled arbitrarily (left/right/center) depending on the dataset.

**Won't do**
- PVDAQ
  - #664 (stale PR)
  - #663
- #1371

**Comment by AdamRJensen** [comment]
# Design spec (DRAFT)
This is a draft for an iotools design specification (this is already partly covered in https://doi.org/10.1016/j.solener.2023.112092)

## Function pattern
Should return a tuple of two elements (data, metadata). The data object is often a pandas DataFrame and the metadata object is a dictionary. If the function does not parse any metadata, an empty dictionary should be returned to follow the standard format.

## Function naming
The main iotools functions should follow the naming patterns `read_x` and `get_x`, where x is a short string describing the data source. For example, `read_bsrn` and `get_bsrn`. 

### Location
For measurement data, the location should be specified using the `station` parameters which is often an abbreviation of the station name. For gridded irradiance databases

## Time
The timeframe of the requested data should be specified using the `start` and `end` parameters. It is highly beneficial that string arguments are also accepted, e.g., '2021-01-01' which can be achieved by using pd.to_datetime. Therefore, the first two lines of an iotools function are typically:

    start = pd.to_datetime(start)
    end = pd.to_datetime(end)

## Credentials
The parameters used to specify credentials depend on the specific service. A suitable combination of the following parameters are recommended `username`, `password`, `api_key`.

## Variable mapping
Functions should by default map variable names to the standard pvlib names listed in the [variable and Symbols](https://github.com/pvlib/pvlib-python/blob/1d854e5a83543de34337a63356efb9c8dd3d859a/pvlib/data/variables_style_rules.csv#L4) list. The parameter determining whether variables are mapped should be named `map_variables`

## Other
Other things to consider: `url`, order of parameters, unit conversion, time_step

**Comment by mikofski** [comment]
Can we also distinguish between a “TMY” style file with 8760 hours of data versus other periods?

And I’d also like to see `coerce_year` be a common feature for TMY style files.

one issue I have with PVGIS is that it’s always UTC which means that it almost always starts 5 to 8 hours before midnight, so another feature I’d like to see is to rotate the data so it starts at midnight in the local timeline

**Reply by AdamRJensen** [reply]
@mikofski can you elaborate on what you mean by distinguishing between tmy and non-tmy data? Are you thinking of how functions are named or something else?

I agree with the `coerce_year` option

I don't think that PVGIS provides any timezone information, so I think that'll have to be left to the user. An Example Gallery of how to localize data to a timezone based on latitude/longitude may be somewhat of a solution.

**Reply by mikofski** [reply]
A TMY file, I believe, starts with the interval from midnight to 1am on Jan 1. Localizing the timezone is only a half solution.  In North America, the pvgis file is still stuck starting at 4pm to 7pm on Dec 31 the previous day, and ends before the 11pm -12am interval on Dec 31 of the following year. We need to rotate the dataframe to make it comparable to other TMY files. All of the intervals should be in the coerced year.

**Reply by adriesse** [reply]
@mikofski I understand your point as I've dealt with this before.  Have you raised the issue with PVGIS?  Does other software do this "rotation" transparently?  Anyway, ```np.roll()``` on the values should do the trick.

**Reply by mikofski** [reply]
Exactly, I’m suggesting `roll_tz` as part of the *standard* API if given would roll the pvgis file so that in the given local tz, it now matches a *standard* TMY

**Reply by AdamRJensen** [reply]
> Exactly, I’m suggesting `roll_tz` as part of the _standard_ API if given would roll the pvgis file so that in the given local tz, it now matches a _standard_ TMY

If we have a ``roll_tz`` parameter, we probably first need to have an `infer_tz` parameter to set a meaningful time zone (can be inferred from the longitude). Agree?

**Reply by adriesse** [reply]
> Exactly, I’m suggesting roll_tz as part of the standard API if given would roll the pvgis file so that in the given local tz, it now matches a standard TMY

I don't know, if it's just fixing a current deficiency of pvgis then it doesn't seem like a thing for a standard API.

**Comment by adriesse** [comment]
There are two general direction iotools could go:

1. Make all data sources look the same to the user, which is most convenient but may hide unique strengths and features of individual sources.
2. Just get the available data into a dataframe (or similar) and let the user take it from there.

I suppose the final product will be somewhere in between, but it may be helpful to keep those extremes in mind and to consider where iotools' aspirations lie in relation to them.

**Comment by echedey-ls** [comment]
Regarding making pvlib.iotools a package (conversation at #2349 and last quarterly meeting in Feb), I think it would be great to start by discussing design criteria for the new pvlib-iotools.

# Design criteria for making iotools a stand-alone package
- Naming of the project (not the package)
  -
  - I propose `pvlib-iotools`
  - Other alternatives:
	- `pviotools`
	- Ideas welcome
- Package import name
  -
  - Same as project name?
  - There's also this namespace package thing, which we can make almost backwards compatible, but makes differentiating the source of code a little bit more difficult
    - In essence, let's say you install with `pip install pvlib-iotools`
    - And use the package like it were a submodule of pvlib: `import pvlib; data, metadata = pvlib.iotools.parse_...(...)`
    - I like this idea so that `pvlib` is like an extendable core of functions, and users opt-in for utilities; and an exception is raised if not found. `pvlib` would need to hold a stub `__init__.py` inside its `iotools` in case we want to warn about not installed capabilities.
  - As a complete different package. What we all know and are familiar with. If project name has hyphens, switch them to underscores.
- First stable version
	-
	- `v0.1.0`? I lean toward this.
	- Or start with the version of pvlib which removes it?
- CI
	-
	- PRs + Pushes to main (as currently is, well, without the label limitation)
	- Run once every two weeks
		- GH docs for that
			- https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#onschedule
			- https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#schedule
			- I don't consider that necessary, but could be useful

Feel free to add to this thread if you find anything unnecessary or anything requires extra attention.

**Comment by AdamRJensen** [comment]
The vision I had for pvlib iotools, which has been tracked in this discussion, is now largely completed. The functions have been standardized and with the next release, pvlib provides access to all major providers of historical irradiance data. This includes commercial and public sources and a range of different types of sources (reanalysis, satellite-derived, and measurements). I therefore do not expect to dedicate significant efforts in the future to adding new iotools.

With that said, at the last IEA PVPS Task 16 meeting in October 2025, several people brought up the desire to have iotools capable of retrieving forecast data (e.g., ECMWF's IFS). Given that we deprecated pvlib's forecast module some years ago, I initially rejected this idea, arguing this should be in its own separate package. However, the iotools module already has some forecasting functions (from the commercial data providers). Therefore, I would like to hear what are the opinions of the @pvlib/pvlib-core, @williamhobbs, and users at large - should pvlib iotools support forecast data? Particularly, @wholmgren can shed some light on the deprecation of the forecast module.

**Reply by williamhobbs** [reply]
> should pvlib iotools support forecast data?

Great question!

I'm leaning towards a "no" on this, because it could be too open-ended. You can look at the significant size, complexity, and maintenance of a package like https://github.com/blaylockbk/Herbie to get an idea of where this could go. 

_Maybe_, if the maintainers and community decided that pvlib's forecast functions would be firmly limited to one or two models, like ECMWF IFS and/or NOAA GFS, then the scope could be manageable, but I suspect there would be frequent requests to add new models, e.g., NOAA HRRR, NOAA RRFS when it comes out, the ensembles versions of GFS and IFS, the AI versions of GFS and IFS, and on and on. And many models have unique "features" related to time intervals (e.g., hourly, 3-hourly, etc.), summary times (avg. vs instantaneous), lag between initialization and outputs being available, etc., that complicate this.

Another option would be to write an iotools function to interface with https://open-meteo.com. It seems to have almost all of the major NWP/AIWP models (go to https://open-meteo.com/en/docs and expand the "Weather Models" section). It applies some sort of interpolation to get hourly data from models with longer intervals (e.g., 3-hourly), and it's open-source. **_It is not free for commercial use_**, though.

**Reply by wholmgren** [reply]
Some thought fragments:

* I have conflict of interest since I'm now in a software products part of DNV working directly with the Solcast team instead of in DNV's advisory/consulting services (yes, there is a big difference)
* Most of the `forecast.py` deprecation discussion is in the relatively short https://github.com/pvlib/pvlib-python/issues/1057 issue. 
* The module was originally proposed in #86 (a full decade ago (gasp!)). It was woefully underdiscussed at the time but even then there was a thought to put it in a separate package. I should've listened!
* I haven't kept up with the time series API offerings from ECMWF, NOAA, Unidata, etc.
* I still think parsing gridded data into the relevant PV time series is far outside of pvlib's scope.
* There is a lot of commercial activity in this space. Some of those vendors provide free or low cost access to researchers.
* Do you want to maintain access to something for free while others are paid to do the same? (the products and reliability are not the same in reality but often approximately the same as far as pvlib's users are concerned)
* `forecast.py` seemed to attract of a lot of users to pvlib, but....
* There was a lot of low quality user support associated with the forecast module. And did any of those users stick around?
* You need a substantial archive of data if you want to do most serious forecast research. That is a super hard problem and way outside of pvlib's scope. This is a fundamental difference with the historical observation data.
* Who are the users? What to they want to do? Will pvlib.iotools actually address their use case? There is too much hassle and noise just to make tinkering a bit easier than they would be able to do with their own code + pvlib.
* There's a lot of great stuff in solarforecastarbiter and it's unfortunate that it hasn't gained more traction. In particular, it provides accessible but required formalisms around what exactly you mean by a forecast. And, optionally, how exactly you intend to compare it to observations. I doubt any of that fits into pvlib.iotools. The current need to work with gridded data and the dependencies is definitely a hurdle. But the code is fairly modular and one could add time series API retrieval options to that package. But I'm guessing all of today's pvlib maintainers would say "nah, not my job" and then I think we're back at the problem of none of us actually want to maintain forecast functionality with the level of tender loving care that it requires.

**Reply by williamhobbs** [reply]
Great thoughts, @wholmgren.

> (a full decade ago (gasp!))

double gasp!

> * You need a substantial archive of data if you want to do most serious forecast research.

This is at least better than it used to be, now that NOAA's [NODD program](https://www.noaa.gov/information-technology/open-data-dissemination) uses AWS/Google/MS for distributing and archiving model outputs. There's still the problem of needing to download full domain grib files when you may only be interested in a single point, so it's still a pretty hard problem.

**Reply by williamhobbs** [reply]
For reference, here are some workflows I currently use for getting forecasts with [Herbie](https://github.com/blaylockbk/Herbie) into a format that works for me with pvlib:

https://github.com/williamhobbs/energy-forecasting-tools, and [forecast_solar.py](https://github.com/williamhobbs/energy-forecasting-tools/blob/main/forecast_solar.py) plus [eft_utilities.py](https://github.com/williamhobbs/energy-forecasting-tools/blob/main/eft_utilities.py) in particular. 

It's not that much code, but using Herbie gives a huge head start.

**Reply by williamhobbs** [reply]
There's also https://dynamical.org/ (https://github.com/dynamical-org). It seems well organized, everything is in zarr format (so getting one or a small number of points is much easier than having to download huge regional/global grib files), and it's free for anyone. Currently includes GFS, GEFS, and HRRR, but ECMWF models are on the roadmap. I know someone on the team there, if anyone wants to connect with them.

**Reply by echedey-ls** [reply]
> The vision I had for pvlib iotools, which has been tracked in this discussion, is now largely completed.

Congrats!! 🚀 

> should pvlib iotools support forecast data? 

I see almost no harm, except for maintenance. Main problem is finding contributors willing to add to it and handling the lifecycle of the project and APIs. I stand with a pvlib-iotools namespace package under pvlib, to keep the releases agile and independent ([comment from time ago](https://github.com/pvlib/pvlib-python/discussions/1528#discussioncomment-13211554)).

**Reply by adriesse** [reply]
I think we have a great feature set and _of course_ there will always be requests for more. I wouldn't say yes or no immediately, but perhaps throw the ball back at the requester to find out more about what they really want. While at first glance it might seem very similar to what io-tools provides: time-stamped irradiance (and more of course); but in fact, there can be multiple additional dimensions. 

Nevertheless, maybe there are one or two use-cases that can be supported more easily than others. Maybe it's just an interface to another system (like open-meteo) with some reformatting of the output for those use cases.
